File: 706.design-hash-map.cpp
/*
 * @lc app=leetcode id=706 lang=cpp
 *
 * [706] Design HashMap
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class MyHashMap {
 public:
  /** Initialize your data structure here. */
  MyHashMap() {}

  /** value will always be non-negative. */
  void put(int key, int value) {
    auto& row = map_[key % kHashSize];
    for (auto& pa : row) {
      if (pa.first == key) {
        pa.second = value;
        return;
      }
    }
    row.push_back({key, value});
  }

  /** Returns the value to which the specified key is mapped, or -1 if this map
   * contains no mapping for the key */
  int get(int key) {
    auto& row = map_[key % kHashSize];
    for (const auto& pa : row) {
      if (pa.first == key) {
        return pa.second;
      }
    }
    return -1;
  }

  /** Removes the mapping of the specified value key if this map contains a
   * mapping for the key */
  void remove(int key) {
    auto& row = map_[key % kHashSize];
    for (auto it = row.begin(); it != row.end(); it++) {
      if (it->first == key) {
        row.erase(it);
        return;
      }
    }
  }

 private:
  static const int kHashSize = 100;
  std::vector<std::pair<int, int>> map_[kHashSize];
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */
// @lc code=end

File: 706.design-hash-map.rs
/*
 * @lc app=leetcode id=706 lang=rust
 *
 * [706] Design HashMap
 */

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
use std::collections::HashMap;
struct MyHashMap {
    map: HashMap<i32, i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyHashMap {
    fn new() -> Self {
        MyHashMap {
            map: HashMap::<i32, i32>::new(),
        }
    }

    fn put(&mut self, key: i32, value: i32) {
        self.map.insert(key, value);
    }

    fn get(&self, key: i32) -> i32 {
        match self.map.get(&key) {
            Some(n) => *n,
            _ => -1,
        }
    }

    fn remove(&mut self, key: i32) {
        self.map.remove(&key);
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * let obj = MyHashMap::new();
 * obj.put(key, value);
 * let ret_2: i32 = obj.get(key);
 * obj.remove(key);
 */
// @lc code=end

#[test]
fn test1() {
    let mut obj = MyHashMap::new();
    obj.put(1, 1);
    obj.put(2, 2);

    assert_eq!(obj.get(1), 1);
    assert_eq!(obj.get(3), -1);
    obj.put(2, 1);
    assert_eq!(obj.get(2), 1);
    obj.remove(2);
    assert_eq!(obj.get(2), -1);
}
